SwiftDungeon/GameManager.swift

```diff
@@ -0,0 +1,165 @@
+//
+//  GameManager.swift
+//  SwiftDungeon
+//
+//  Created by Toro Juan D. on 3/12/18.
+//  Copyright © 2018 Toro Juan D. All rights reserved.
+//
+
+import Foundation
+import GameKit
+
+class GameManager {
+    
+    // Map
+    private let map = Map()
+    
+    // Entities
+    let player = Player()
+    var enemies:[Enemy] = []
+    private var maxEnemies = 3
+    private let enemyFactory = EnemyFactory()
+    
+    // Loading check
+    var loading: Bool = false
+    
+    // UI
+    var joystick: Joystick?
+    var attackButton: AttackButton?
+    
+    // Scene
+    weak var scene: GameScene?
+    
+    func startGame(size: CGSize) {
+        // Make Map
+        map.zPosition = -1
+        map.position = CGPoint(x: size.width / 2, y: size.height / 2 + 120)
+        map.setScale(2.5)
+        
+        // Make Player
+        scene?.addChild(map)
+        player.zPosition = 1
+        player.position = map.playerSpawn
+        player.setScale(5)
+        scene?.addChild(player)
+        
+        //Init the Joystick
+        joystick = Joystick()
+        joystick?.scene = scene
+        joystick?.start(size: size)
+        
+        //Init the Attack Button
+        attackButton = AttackButton()
+        attackButton?.scene = scene
+        attackButton?.start(size: size)
+        
+        // Populate level with enemies
+        populateEnemies()
+    }
+    
+    func update(_ currentTime: TimeInterval) {
+        
+        // Update Player
+        player.update(currentTime)
+        player.setDirection(dir: (joystick?.dirVector)!)
+        
+        // Keep player within level bounds
+        if(player.position.y < 600)
+        {
+            player.position.y = CGFloat(600)
+        }
+        if(player.position.x < 0)
+        {
+            player.position.x = CGFloat(0)
+        }
+        if(player.position.y > (scene?.size.height)! - 200)
+        {
+            player.position.y = CGFloat((scene?.size.height)! - 200)
+        }
+        if(player.position.x > (scene?.size.width)!)
+        {
+            player.position.x = CGFloat((scene?.size.width)!)
+        }
+        
+        // Update Enemies and check for collisions
+        var enemiesToBeDeleted: [Int] = []
+        
+        for enemy in (enemies) {
+            if(!loading) {
+                enemy.update(currentTime)
+            }
+            enemy.setTarget(player.position)
+            
+            let playerCollision = enemy.collision(items: [player]).first
+            
+            if let _ = playerCollision  {
+                if player.isAttacking {
+                    enemy.removeFromParent()
+                    enemiesToBeDeleted.append(enemies.index(of: enemy)!)
+                }
+                else {
+                    player.takeDamage()
+                    print(player.health)
+                }
+            }
+        }
+        
+        deleteEnemies(enemiesToBeDeleted)
```

Comment:
All these delete code is only really necessary because we keep track of the enemies in multiple places. 1) they are attached to the scene node, 2) they are stored in an array inside GameScene. Like the example project
Essentially if we left them only with #1 deleting them would be a lot easier. Simply call .removeFromParent() but we would have a harder time finding them because we would have to traverse the SceneGraph.